### PR TITLE
PP 49 - Unify tests

### PR DESCRIPTION
**Pull request by antomor**
Remove code duplication and unify test among CI and local execution.
Create some bash utility functions to be called by external scripts.

### CODE REVIEW

File: docker/docker-compose.yml [hunk 1]
```diff
@@ -12,6 +12,7 @@ services:
         command: --regtest
         ports:
             - '127.0.0.1:4444:4444'
+            - '127.0.0.1:4445:4445'
```
Comment (antomor) [inline docker/docker-compose.yml:15]: This change is already included in an open PR: https://github.com/rsksmart/rif-relay/pull/158/files#diff-423deb13b7c401b1a7f41ee91c77f722e11d2f317d6a66b546524e8a04cc8b03R15

**Review by wilsoniovlabs — APPROVED** [review]
LGTM

**Comment by antomor** [comment]
@mortelli here is the PP-49 PR. Do you mind reviewing it?

**Review by mortelli — COMMENTED** [review]
so i ran the tests on my machine and they seem to work OK :+1: and the CI also seems to be working fine, so i think we're pretty much ready to merge.

i have a couple more comments besides the ones down below:
1. i know you didn't introduce this change, but there seems to be 6 groups in the (original) CI and 5 in the docker version. what's your take on this? i know that they are consistent now, but maybe we want to go back to 6 instead of 5?
    1. weirdly, `relayclient/RelaySelectionManager.test.ts` seems to be on both groups 2 and 6.
2. the only difference between the CI and the docker tests that i can see right now is the set-up and tear down. do you think the copying and removal of folders is not necessary in the docker tests?
3. a minor gripe: we could add names to the CI threads that lack them. if you check out for example thread 2, you'll see that it is named after its execution command. not that `Tests without DB reset Set X` is terribly helpful, but it's at least more readable :shrug: 

thanks for this good solution, ready to approve soon :+1:

File: .circleci/config.yml [hunk 2]
```diff
@@ -102,147 +102,58 @@ jobs: # a collection of steps
...
-                      npx truffle test --network regtest test/common/VersionManager.test.ts
-                      npx truffle test --network regtest test/regressions/PayableWithEmit.test.ts
-                      npx truffle test --network regtest test/relayclient/AccountManager.test.ts
-                      npx truffle test --network regtest test/relayclient/ContractInteractor.test.ts
-                      npx truffle test --network regtest test/relayclient/Configurator.test.ts
-                      rm -rf contracts
-                      rm -rf migrations
-                      rm -rf contract-addresses.json
-            - store_artifacts:
-                  path: ~/gls/logs
```
[the remark this answers is not in the document]
Comment (antomor) [reply .circleci/config.yml:127]: Good catch, I think that logs could be useful in case of failing, thanks!

**Comment by antomor** [comment]
@mortelli thanks for the comments, they make sense for me.

> i know you didn't introduce this change, but there seems to be 6 groups in the (original) CI and 5 in the docker version. what's your take on this? i know that they are consistent now, but maybe we want to go back to 6 instead of 5?

I thought there was a legitimate reason for having many groups, but after merging some of them, I noticed that the tests run successfully. My questionable opinion is that I wouldn't split tests unless for a specific reason; Ideally, I wouldn't specify the test order at all, but I suppose it wouldn't work in this case. However, if we want to split them, no problem for me, let's discuss the rationale behind the split and we could add some comments as a reference.

> weirdly, relayclient/RelaySelectionManager.test.ts seems to be on both groups 2 and 6.

Thanks for pointing it out :-)

>the only difference between the CI and the docker tests that i can see right now is the set-up and tear down. do you think the copying and removal of folders is not necessary in the docker tests?

Sorry, I can't get what you mean. I think the contracts and migrations folder are copied within the `run_batch` function that is actually used both on docker and on CI. I am sure I am missing something...

>a minor gripe: we could add names to the CI threads that lack them. if you check out for example thread 2, you'll see that it is named after its execution command. not that Tests without DB reset Set X is terribly helpful, but it's at least more readable 🤷

Not so minor, I agree with you, but I don't have any idea about the names to set on those test suites. What would you suggest?

**Comment by mortelli** [comment]
> I thought there was a legitimate reason for having many groups, but after merging some of them, I noticed that the tests run successfully. My questionable opinion is that I wouldn't split tests unless for a specific reason; Ideally, I wouldn't specify the test order at all, but I suppose it wouldn't work in this case. However, if we want to split them, no problem for me, let's discuss the rationale behind the split and we could add some comments as a reference.

you're right. the main issue here is that we are not familiar enough with the intricacies of the test suite to confidently state why certain ones should be executed sequentially and others in parallel, for example. ignorance is a rather bad excuse, but until we have better ownership of these files my strategy would be "keep as is until we find a reason not to" (like a bug).

having said that, we were already executing 5 threads locally and 6 on the CI, so we have precedent to have 5 groups instead of 6. so it's fine either way, we can leave it as it is right now.
 
> Sorry, I can't get what you mean. I think the contracts and migrations folder are copied within the `run_batch` function that is actually used both on docker and on CI. I am sure I am missing something...

what i meant by this was that the CI threads have these lines:

```bash
cp -r node_modules/@rsksmart/rif-relay-contracts/contracts .
cp -r node_modules/@rsksmart/rif-relay-contracts/migrations .
...
rm -rf contracts
rm -rf migrations
rm -rf contract-addresses.json
```

which i believe are not present in the docker functions. but maybe we don't need them since we're doing docker-compose `up` and `down`?
 
> Not so minor, I agree with you, but I don't have any idea about the names to set on those test suites. What would you suggest?

at this point, i would just go with `Test thread #n` for `n=1` through `5`

**Review by mortelli — APPROVED** [review]

**Comment by mortelli** [comment]
> what i meant by this was that the CI threads have these lines:
> 
> ```shell
> cp -r node_modules/@rsksmart/rif-relay-contracts/contracts .
> cp -r node_modules/@rsksmart/rif-relay-contracts/migrations .
> ...
> rm -rf contracts
> rm -rf migrations
> rm -rf contract-addresses.json
> ```
> 
> which i believe are not present in the docker functions. but maybe we don't need them since we're doing docker-compose `up` and `down`?

never mind, it seems i missed them. consider this solved, apologies

**Pull request merged**